### DISCUSSION TITLE
Does it download S3 files locally or can query them directly on S3?

### DISCUSSION THREAD

**Discussion by collimarco** [openobserve/openobserve#876] [General]
I am wondering how OpenObserve works. 

I know that Apache Arrow and Parquet can use condition pushdown to load only a partial subset of rows from a remote file. However S3 latency seems very high.

Does OpenObserve download the S3 files locally and then query them?
Or it can query them directly on S3 without downloading them entirely (i.e. download only the rows that match a condition)?

I am considering using/contributing to this project but I don't understand exactly the approach.

**Comment by prabhatsharma** [comment]
We download only the necessary files from S3, provided that the partition exists. By default, the partition is based on the Year, Month, Day, and Hour. Additional partitions such as 'host', 'status', etc., can be enabled based on fields in the data. If you have three days' worth of data in S3 and you only query for data from the last hour, we will only download the data from the last hour from S3.

Additionally, we maintain an in-memory cache that allows us to execute queries using data in memory, thereby eliminating the need to access S3 for the same data.

**Reply by collimarco** [reply]
Thanks for the clarification! 

Why do you say "in-memory cache", don't you keep a copy of the S3 files on local disk? 

Also, have you discarded the idea of using "condition pushdown" to read only parts of the files on S3? https://arrow.apache.org/docs/python/dataset.html#filtering-data

**Reply by prabhatsharma** [reply]
No, We don't keep a copy of the data locally on disk. We have not discarded the idea of pulling only required data from a file. We just have not been able to get there. We got more important items on the roadmap for now.
